projects/vdk-core/src/vdk/internal/core/errors.py

```diff
@@ -13,25 +13,107 @@
+        error_message = ""
+        if error_message_lines and isinstance(error_message_lines[0], ErrorMessage):
+            message = error_message_lines[0]
+            error_message += message.summary
+            error_message += message.what
+            error_message += message.why
+            error_message += message.consequences
+            error_message += message.countermeasures
+        elif error_message_lines and isinstance(error_message_lines[0], dict):
+            message = error_message_lines[0]
+            error_message += "\n".join(message.values())
+        else:
+            for line in error_message_lines:
+                error_message += line
+                error_message += "\n"
+        super().__init__(error_message)
+        self._vdk_resolvable_actual = vdk_resolvable_actual
+        self._vdk_type = vdk_type
+        self._prettify_message(str(error_message))
+
+    def __str__(self):
+        return self._pretty_message
+
+    def __repr__(self):
+        return self._pretty_message
+
+    def _prettify_message(self, message):
+        box_char = "+"
+        header = f"{self.__class__.__name__}: An error of type {self._vdk_type} occurred. Error should be fixed by {self._vdk_resolvable_actual}"
+        lines = message.split("\n")
+        max_len = len(header)
+        # Wrap the message lines
+        wrapped_lines = []
+        for line in lines:
+            # Find the nearest space if a line is
+            # longer than the header and wrap around it
+            # Do this until the remainder of the line is shorter
+            # than the header
+            if len(line) > max_len:
+                l = line
+                while len(l) > max_len:
+                    br = max_len
+                    while l[br] != " " and br > 0:
+                        br -= 1
+                    if br == 0:
+                        br = max_len
+                    wrapped_lines.append(l[: br + 1])
+                    l = l[br + 1 :]
+                wrapped_lines.append(l)
+            else:
+                wrapped_lines.append(line)
+        # build th header
+        header = box_char + header.center(max_len + 6) + box_char
+        # build the lines with the box char
+        lines = [
+            box_char + "  " + s.ljust(max_len + 4) + box_char for s in wrapped_lines
+        ]
+        # build the box sides
+        side = (max_len + 8) * box_char
+        self._pretty_message = "\n".join(
+            ["\n" + side, header, side, "\n".join(lines), side]
+        )
+
+
+class SkipRemainingStepsException(BaseVdkError):
+    """
+    An exception used to skip the remaining steps of a Data Job.
+
+    When this exception is thrown from a data job python step, all the other steps will
+    be skipped and the data job execution will exit and be marked as success.
+    """
+
+    def __init__(self, *error_message_lines: str):
+        super().__init__(None, None, *error_message_lines)
+
+
+# RESOLVABLE CONTEXT
```

Comment:
IMO, we should consider refactoring the  file for clearer code organisation. Currently, it contains both error definitions and the logic for error resolution and responsibility assignment.
IMO, keeping only error definitions and related utilities as 'report and rethrow' would make the module a lot easier to read and understand. We can move the error-resolving logic and classes to a new module  named 'error_resolvers.py' or something similar.
By doing this separation, we will make our codebase more intuitive, navigable (tbh reading the file was pretty hard), and maintainable, ensuring that each module has a single, clear responsibility, which will be beneficial for the development team in the long run.

PS: We might not be doing it because of compatibility issues, I am not sure but i do think maybe later on this should be changed

Reply:
I was thinking the same thing initially, but there's a lot of `errors.ResolvableBy` and `errors.log_and_throw` and similar statements throughout plugin code, so I decided against it. I'll add it to https://github.com/vmware/versatile-data-kit/issues/2680 as a suggestion, or open a separate ticket.